File: native/ops/point.cpp
#include "ops/point.h"

#include <stdexcept>

#include "backends/cpu/point.h"

namespace mlx_lattice {

namespace {

// MARK: - validation

void validate_coords(const mx::array& coords) {
    if (coords.ndim() != 2 || coords.shape(1) != 4) {
        throw std::invalid_argument("coords must have shape (N, 4).");
    }
    if (coords.dtype() != mx::int32 && coords.dtype() != mx::int64) {
        throw std::invalid_argument("coords must be int32 or int64.");
    }
}

void validate_positive(Triple values, const char* name) {
    for (auto value : values) {
        if (value <= 0) {
            throw std::invalid_argument(
                std::string(name) + " values must be positive."
            );
        }
    }
}

} // namespace

// MARK: - helpers

std::vector<Triple> kernel_offsets(Triple kernel_size) {
    validate_positive(kernel_size, "kernel_size");

    std::array<std::vector<int>, 3> axes;
    for (int axis = 0; axis < 3; ++axis) {
        int size = kernel_size[axis];
        if (size % 2 == 1) {
            int radius = size / 2;
            for (int value = -radius; value <= radius; ++value) {
                axes[axis].push_back(value);
            }
        } else {
            for (int value = 0; value < size; ++value) {
                axes[axis].push_back(value);
            }
        }
    }

    std::vector<Triple> offsets;
    offsets.reserve(axes[0].size() * axes[1].size() * axes[2].size());
    for (auto x : axes[0]) {
        for (auto y : axes[1]) {
            for (auto z : axes[2]) {
                offsets.push_back({x, y, z});
            }
        }
    }
    return offsets;
}

// MARK: - api

mx::array downsample_coords(const mx::array& coords, Triple stride) {
    validate_coords(coords);
    validate_positive(stride, "stride");
    return cpu::downsample_coords(coords, stride);
}

KernelMapData
build_kernel_map(const mx::array& coords, Triple kernel_size, Triple stride) {
    validate_coords(coords);
    validate_positive(stride, "stride");
    return cpu::build_kernel_map(coords, kernel_size, stride);
}

} // namespace mlx_lattice

File: src/mlx_lattice/point.py
from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass
from typing import cast

import mlx.core as mx

from mlx_lattice._native import (
    build_generative_map as _build_generative_map,
)
from mlx_lattice._native import (
    build_kernel_map as _build_kernel_map,
)
from mlx_lattice._native import (
    downsample_coords as _downsample_coords,
)
from mlx_lattice.types import Triple, triple


@dataclass(frozen=True)
class KernelMap:
    maps: mx.array
    sizes: mx.array
    kernels: mx.array
    residual_maps: mx.array
    residual_kernels: mx.array
    residual_offsets: mx.array
    out_coords: mx.array
    offsets: tuple[Triple, ...]

    @property
    def center(self) -> int:
        try:
            return self.offsets.index((0, 0, 0))
        except ValueError:
            return -1


def kernel_offsets(kernel_size: int | Sequence[int]) -> tuple[Triple, ...]:
    axes = []
    for size in triple(kernel_size, name='kernel_size'):
        if size <= 0:
            raise ValueError('kernel_size values must be positive.')
        if size % 2:
            radius = size // 2
            axes.append(range(-radius, radius + 1))
        else:
            axes.append(range(size))
    return tuple(
        (int(x), int(y), int(z))
        for x in axes[0]
        for y in axes[1]
        for z in axes[2]
    )


def downsample(
    coords: mx.array,
    stride: int | Sequence[int] = 2,
) -> mx.array:
    if coords.ndim != 2 or coords.shape[1] != 4:
        raise ValueError('coords must have shape (N, 4).')
    if coords.dtype not in (mx.int32, mx.int64):
        raise ValueError('coords must be int32 or int64.')

    step = triple(stride, name='stride')
    if any(value <= 0 for value in step):
        raise ValueError('stride values must be positive.')
    return _downsample_coords(coords, step)


def build_kernel_map(
    coords: mx.array,
    kernel_size: int | Sequence[int] = 3,
    stride: int | Sequence[int] = 1,
    padding: int | Sequence[int] = 0,
) -> KernelMap:
    if coords.ndim != 2 or coords.shape[1] != 4:
        raise ValueError('coords must have shape (N, 4).')
    if coords.dtype not in (mx.int32, mx.int64):
        raise ValueError('coords must be int32 or int64.')

    kernel = triple(kernel_size, name='kernel_size')
    step = triple(stride, name='stride')
    pad = triple(padding, name='padding')
    if any(value <= 0 for value in step):
        raise ValueError('stride values must be positive.')
    if any(value < 0 for value in pad):
        raise ValueError('padding values must be non-negative.')

    (
        maps,
        sizes,
        kernels,
        residual_maps,
        residual_kernels,
        residual_offsets,
        out_coords,
        offset_values,
    ) = _build_kernel_map(coords, kernel, step, pad)
    offsets = _offsets_from_array(offset_values)
    (
        maps,
        sizes,
        kernels,
        residual_maps,
        residual_kernels,
        residual_offsets,
        out_coords,
    ) = _place_cached_map_arrays(
        coords,
        maps,
        sizes,
        kernels,
        residual_maps,
        residual_kernels,
        residual_offsets,
        out_coords,
    )
    return KernelMap(
        maps=maps,
        sizes=sizes,
        kernels=kernels,
        residual_maps=residual_maps,
        residual_kernels=residual_kernels,
        residual_offsets=residual_offsets,
        out_coords=out_coords,
        offsets=offsets,
    )


def build_generative_map(
    coords: mx.array,
    kernel_size: int | Sequence[int] = 2,
    stride: int | Sequence[int] = 2,
) -> KernelMap:
    if coords.ndim != 2 or coords.shape[1] != 4:
        raise ValueError('coords must have shape (N, 4).')
    if coords.dtype not in (mx.int32, mx.int64):
        raise ValueError('coords must be int32 or int64.')

    kernel = triple(kernel_size, name='kernel_size')
    step = triple(stride, name='stride')
    if any(value <= 0 for value in step):
        raise ValueError('stride values must be positive.')
    (
        maps,
        sizes,
        kernels,
        residual_maps,
        residual_kernels,
        residual_offsets,
        out_coords,
        offset_values,
    ) = _build_generative_map(coords, kernel, step)
    offsets = _offsets_from_array(offset_values)
    (
        maps,
        sizes,
        kernels,
        residual_maps,
        residual_kernels,
        residual_offsets,
        out_coords,
    ) = _place_cached_map_arrays(
        coords,
        maps,
        sizes,
        kernels,
        residual_maps,
        residual_kernels,
        residual_offsets,
        out_coords,
    )
    return KernelMap(
        maps=maps,
        sizes=sizes,
        kernels=kernels,
        residual_maps=residual_maps,
        residual_kernels=residual_kernels,
        residual_offsets=residual_offsets,
        out_coords=out_coords,
        offsets=offsets,
    )


def _offsets_from_array(values: mx.array) -> tuple[Triple, ...]:
    rows = cast(list[list[int]], values.tolist())
    return tuple((int(row[0]), int(row[1]), int(row[2])) for row in rows)


def _place_cached_map_arrays(
    coords: mx.array,
    *arrays: mx.array,
) -> tuple[mx.array, ...]:
    gpu_available = mx.metal.is_available() or mx.cuda.is_available()
    if (
        coords.dtype == mx.int32
        and gpu_available
        and mx.default_device() == mx.Device(mx.gpu)
    ):
        device = mx.default_device()
        return tuple(
            mx.contiguous(array, stream=device) for array in arrays
        )
    return arrays
